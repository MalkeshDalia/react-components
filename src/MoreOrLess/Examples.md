### Examples

```js
initialState = { expanded: false }

function onExpandedChange() {
  setState({ expanded: !state.expanded })
}

const templateExpanded = () => (
  <FlexView className='card'>
    <FlexView grow shrink className='description'>
      <FlexView column>
        <FlexView className='title'>The Student Hotel</FlexView>
        <FlexView className='subtitle'>
          <FlexView className='author'>Taylor Cambell</FlexView>
          <FlexView className='date'>7th of March</FlexView>
        </FlexView>
        What is boutique hotel design, and why does it
        matter? There are fashion trends in hotel design just
        as there are in other areas. In the late eighties and
        through the nineties, that fashion was for clean, stark,
        minimalist design, particularly in city hotels. This was
        such a diversion from the traditional way.
      </FlexView>
    </FlexView>
    <FlexView grow shrink={false} hAlignContent='right'>
      <img src='moreOrLess.png' />
    </FlexView>
  </FlexView>
);

const templateCollapsed = () => (
  <FlexView className='card' style={{ maxHeight: 116 }}>
    <FlexView className='description'>
      <FlexView column>
        <FlexView className='title'>The Student Hotel</FlexView>
        <FlexView className='subtitle'>
          <FlexView className='author'>Taylor Cambell</FlexView>
          <FlexView className='date'>7th of March</FlexView>
        </FlexView>
      </FlexView>
    </FlexView>
    <FlexView grow hAlignContent='right' vAlignContent='center'>
      <img src='moreOrLess2.png' style={{ width: 'auto', height: 80 }} />
    </FlexView>
  </FlexView>
);


const { expanded } = state;
const icons = { expanded: 'angleUpsvg', collapsed: 'angleDown' };

<MoreOrLess {...{ expanded, onExpandedChange, icons }}>
  { expanded ? templateExpanded() : templateCollapsed()}
</MoreOrLess>
```
