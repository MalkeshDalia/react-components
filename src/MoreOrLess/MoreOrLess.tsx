import * as React from 'react';
import cx = require('classnames');
import { props, t, ReactChildren } from '../utils';
import Icon from '../Icon/Icon';
import FlexView from 'react-flexview';

export type MoreOrLessProps = {
  /** panel content */
  children: any, // TODO: t.ReactChildren
  /** whether the panel should be expanded or not */
  expanded: boolean,
  /** called on toggle */
  onExpandedChange: (expanded: boolean) => void,
  /** icons for expanded and collapsed panel */
  icons: {
    expanded: string,
    collapsed: string
  },
  /** props for wrapper FlexView */
  wrapperProps?: object,
  /** an optional class name to pass to the component */
  className?: string,
  /** an optional style object to pass to the component */
  style?: React.CSSProperties
};

export const Props = {
  children: ReactChildren,
  expanded: t.Boolean,
  onExpandedChange: t.Function,
  icons: t.struct({
    expanded: t.String,
    collapsed: t.String
  }),
  wrapperProps: t.maybe(t.Object),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
};

/**
 * A panel used to alternately display short or long version of the content
 */
@props(Props)
export default class MoreOrLess extends React.PureComponent<MoreOrLessProps> {

  toggleExpanded = () => {
    this.props.onExpandedChange(!this.props.expanded);
  };

  templateExpandButton = (icon: string, toggleExpanded: React.EventHandler<React.SyntheticEvent<HTMLDivElement>>) => {
    return (
      <FlexView
        hAlignContent='center'
        vAlignContent='center'
        className='expand-button'
        onClick={toggleExpanded}
        shrink={false}
      >
        <Icon icon={icon} className='expand-button-icon' />
      </FlexView>
    );
  };

  render() {
    const {
      props: { children, className, style, expanded, icons, wrapperProps },
      toggleExpanded
    } = this;

    const panelState = expanded ? 'more' : 'less';
    const icon = expanded ? icons.expanded : icons.collapsed;

    return (
      <FlexView column shrink={false} {...wrapperProps}
        style={style}
        className={cx('more-or-less', panelState, className)}
      >
        {children}
        {this.templateExpandButton(icon, toggleExpanded)}
      </FlexView>
    );
  }
}
