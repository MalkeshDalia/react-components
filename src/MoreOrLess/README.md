# MoreOrLess

A panel used to alternately display short or long version of the content

## Props
|Name|Type|Default|Description|
|----|----|-------|-----------|
| **children** | <code>ReactChildren</code> |  | **required**. Panel content |
| **expanded** | <code>Boolean</code> |  | **required**. Whether the panel should be expanded or not |
| **onExpandedChange** | <code>Function</code> |  | **required**. Called on toggle |
| **icons** | <code>Struct{expanded: String, collapsed: String}</code> |  | **required**. Icons for expanded and collapsed panel |
| **wrapperProps** | <code>Object</code> |  | *optional*. Props for wrapper FlexView |
| **className** | <code>String</code> |  | *optional*. Additional `className` for wrapper element |
| **style** | <code>Object</code> |  | *optional*. Inline-style overrides for wrapper element |