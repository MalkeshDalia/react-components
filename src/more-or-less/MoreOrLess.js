import React from 'react';
import cx from 'classnames';
import { props, t, skinnable } from '../utils';
import Icon from '../Icon/Icon';
import FlexView from 'react-flexview';

export const Props = {
  children: t.ReactChildren,
  expanded: t.Boolean,
  onExpandedChange: t.Function,
  icons: t.struct({
    expanded: t.String,
    collapsed: t.String
  }),
  wrapperProps: t.maybe(t.Object),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
};

/**
 * A panel used to alternately display short or long version of the content
 * @param children - panel content
 * @param expanded - whether the panel should be expanded or not
 * @param onExpandedChange - called on toggle
 * @param icons - icons for expanded and collapsed panel
 * @param wrapperProps - props for wrapper FlexView
 */
@skinnable()
@props(Props)
export default class MoreOrLess extends React.PureComponent {

  toggleExpanded = () => {
    this.props.onExpandedChange(!this.props.expanded);
  };

  getLocals() {
    const {
      props: { children, className, style, expanded, icons, wrapperProps },
      toggleExpanded
    } = this;

    const panelState = expanded ? 'more' : 'less';
    const icon = expanded ? icons.expanded : icons.collapsed;
    return {
      children,
      icon,
      toggleExpanded,
      wrapperProps: {
        ...wrapperProps,
        style,
        className: cx('more-or-less', panelState, className)
      }
    };
  }

  templateExpandButton = ({ icon, toggleExpanded }) => {
    return (
      <FlexView
        hAlignContent='center'
        vAlignContent='center'
        className='expand-button'
        onClick={toggleExpanded}
        shrink={false}
      >
        <Icon icon={icon} className='expand-button-icon' />
      </FlexView>
    );
  };

  template({ children, wrapperProps, icon, toggleExpanded }) {
    return (
      <FlexView column shrink={false} {...wrapperProps}>
        {children}
        {this.templateExpandButton({ icon, toggleExpanded })}
      </FlexView>
    );
  }
}
