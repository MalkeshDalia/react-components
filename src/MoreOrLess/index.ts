import MoreOrLess, { MoreOrLessProps } from './MoreOrLess';
export default MoreOrLess;
export { MoreOrLessProps };
