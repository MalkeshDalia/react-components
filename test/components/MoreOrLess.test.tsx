import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import MoreOrLess from '../../src/MoreOrLess';

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

const exampleProps = {
  children: 'content',
  className: 'fancy-class-name',
  icons: {
    expanded: 'angle-up',
    collapsed: 'angle-down'
  },
  onExpandedChange: () => {}
};

const componentMore = mount(<MoreOrLess {...exampleProps} expanded={true} />);

const componentLess = mount(<MoreOrLess {...exampleProps} expanded={false} />);


describe('MoreOrLess', () => {

  it('renders correctly when expanded', () => {
    const component = renderer.create(
      <MoreOrLess {...exampleProps} expanded />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders correctly when not expanded', () => {
    const component = renderer.create(
      <MoreOrLess {...exampleProps} expanded={false} />
    );
    expect(component).toMatchSnapshot();
  });

  describe('getLocals', () => {

    it('computes className', () => {
      expect(componentMore.render().find('.more-or-less').hasClass('more fancy-class-name')).toBe(true);
      expect(componentLess.render().find('.more-or-less').hasClass('less fancy-class-name')).toBe(true);
    });

    it('computes icon', () => {
      expect(componentMore.find('.expand-button-icon').hasClass('icon-angle-up')).toBe(true);
      expect(componentLess.find('.expand-button-icon').hasClass('icon-angle-down')).toBe(true);
    });

  });

});
