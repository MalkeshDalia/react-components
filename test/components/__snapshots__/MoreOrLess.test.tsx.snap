// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreOrLess renders correctly when expanded 1`] = `
<div
  className="react-flex-view flex-column more-or-less more fancy-class-name"
  style={
    Object {
      "MozBoxFlex": "0 0 auto",
      "WebkitBoxFlex": "0 0 auto",
      "WebkitFlex": "0 0 auto",
      "flex": "0 0 auto",
      "msFlex": "0 0 auto",
    }
  }
>
  content
  <div
    className="react-flex-view align-content-center justify-content-center expand-button"
    onClick={[Function]}
    style={
      Object {
        "MozBoxFlex": "0 0 auto",
        "WebkitBoxFlex": "0 0 auto",
        "WebkitFlex": "0 0 auto",
        "flex": "0 0 auto",
        "msFlex": "0 0 auto",
      }
    }
  >
    <i
      className="icon icon-angle-up expand-button-icon"
      onClick={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`MoreOrLess renders correctly when not expanded 1`] = `
<div
  className="react-flex-view flex-column more-or-less less fancy-class-name"
  style={
    Object {
      "MozBoxFlex": "0 0 auto",
      "WebkitBoxFlex": "0 0 auto",
      "WebkitFlex": "0 0 auto",
      "flex": "0 0 auto",
      "msFlex": "0 0 auto",
    }
  }
>
  content
  <div
    className="react-flex-view align-content-center justify-content-center expand-button"
    onClick={[Function]}
    style={
      Object {
        "MozBoxFlex": "0 0 auto",
        "WebkitBoxFlex": "0 0 auto",
        "WebkitFlex": "0 0 auto",
        "flex": "0 0 auto",
        "msFlex": "0 0 auto",
      }
    }
  >
    <i
      className="icon icon-angle-down expand-button-icon"
      onClick={[Function]}
      style={
        Object {
          "color": undefined,
        }
      }
    />
  </div>
</div>
`;
